// react-f/lib/src/field/Data.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Data_1 = require("../form/Data");
var AsyncValidation_1 = require("../form/AsyncValidation");
var createHOC_1 = require("../createHOC");
exports.default = createHOC_1.default({
    name: "Field/Data",
    contextTypes: Data_1.default.IDs.concat(AsyncValidation_1.default.IDs),
    class: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = {
                parserFunc: null,
                stringifyFunc: null,
                fieldType: null
            };
            return _this;
        }
        class_1.prototype.parse = function (value) {
            var parserFunc = this.state.parserFunc;
            return parserFunc ? parserFunc(value) : value;
        };
        class_1.prototype.stringify = function (value) {
            var stringifyFunc = this.state.stringifyFunc;
            var stringValue = stringifyFunc ? stringifyFunc(value) : value;
            if (null === stringValue || undefined === stringValue) {
                return "";
            }
            else {
                return stringValue;
            }
        };
        class_1.prototype.getChildProps = function () {
            var _this = this;
            var _a = this.state, parserFunc = _a.parserFunc, stringifyFunc = _a.stringifyFunc;
            var _b = Data_1.default.props(this) || Data_1.default.defaultProps, setValue = _b.setValue, getValue = _b.getValue, updateValue = _b.updateValue;
            var value = getValue(this.props.field);
            var isCheckbox = this.props.type === "checkbox" || this.state.fieldType === "checkbox";
            var isRadio = this.props.type === "radio" || this.state.fieldType === "radio";
            var stringValue = this.stringify(value);
            return {
                value: (isCheckbox || isRadio) ? "" : stringValue,
                checked: isRadio ? stringValue == this.props.value : !!stringValue,
                setValue: function (value) {
                    setValue(_this.props.field, _this.parse(isCheckbox ? !!value : value));
                },
                updateValue: function (updater) {
                    updateValue(_this.props.field, updater);
                },
                onChange: function (e) {
                    var value = isCheckbox
                        ? e.currentTarget.checked
                        : e.currentTarget.value;
                    setValue(_this.props.field, _this.parse(value));
                },
                onParse: function (parserFunc) { return _this.setState({ parserFunc: parserFunc }); },
                onStringify: function (stringifyFunc) { return _this.setState({ stringifyFunc: stringifyFunc }); },
                setFieldType: function (fieldType) { return _this.setState({ fieldType: fieldType }); }
            };
        };
        return class_1;
    }(React.Component))
});
