// react-f/lib/src/field/Submitable.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var createHOC_1 = require("../createHOC");
var Submitable_1 = require("../form/Submitable");
exports.default = createHOC_1.default({
    name: "Field/Submitable",
    contextTypes: Submitable_1.default.IDs,
    class: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.getChildProps = function () {
            var _a = Submitable_1.default.props(this) || Submitable_1.default.defaultProps, submitted = _a.submitted, submitting = _a.submitting;
            return {
                submitted: submitted,
                submitting: submitting
            };
        };
        return class_1;
    }(React.Component))
});
