// react-f/lib/src/form/Field.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Data_1 = require("./Data");
var createHOC_1 = require("../createHOC");
exports.default = createHOC_1.default({
    name: "Form/Fieldprovider",
    defaultChildProps: {
        field: function () { return ({}); }
    },
    class: /** @class */ (function (_super) {
        __extends(FormWrapper, _super);
        function FormWrapper() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = {
                data: {}
            };
            return _this;
        }
        FormWrapper.prototype.getChildProps = function () {
            var _a = Data_1.default.props(this) || Data_1.default.defaultProps, getValue = _a.getValue, setValue = _a.setValue;
            return {
                field: function (name) { return ({
                    value: getValue(name) || "",
                    onChange: function (e) { return setValue(name, e.currentTarget.value); }
                }); }
            };
        };
        return FormWrapper;
    }(React.Component))
});
