// react-f/lib/src/form/Touchable.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Objects_1 = require("../Objects");
var createHOC_1 = require("../createHOC");
exports.default = createHOC_1.default({
    name: "Form/Touchable",
    class: /** @class */ (function (_super) {
        __extends(FormWrapper, _super);
        function FormWrapper() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = {
                touched: {}
            };
            _this.touched = function (field) {
                return Objects_1.get(_this.state.touched, field) || false;
            };
            _this.touch = function (field) {
                _this.setState(function (s) { return ({
                    touched: Objects_1.set(s.touched, field, true)
                }); });
            };
            return _this;
        }
        FormWrapper.prototype.getChildProps = function () {
            return {
                touched: this.touched,
                touch: this.touch
            };
        };
        return FormWrapper;
    }(React.Component))
});
