// react-f/lib/src/form/Wizard.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var createHOC_1 = require("../createHOC");
exports.default = createHOC_1.default({
    name: "Other/Wizard",
    defaultChildProps: {
        step: 0,
        nextStep: function () { return void 0; },
        previousStep: function () { return void 0; },
        setStep: function () { return void 0; }
    },
    class: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = {
                step: 0,
            };
            return _this;
        }
        class_1.prototype.getChildProps = function () {
            var _this = this;
            return {
                step: this.state.step,
                nextStep: function () { return _this.setState({ step: _this.state.step + 1 }); },
                previousStep: function () { return _this.setState({ step: _this.state.step - 1 }); },
                setStep: function (step) { return _this.setState({ step: step }); }
            };
        };
        return class_1;
    }(React.Component))
});
